src/test/java/org/sagebionetworks/workers/util/semaphore/ReadLockImplTest.java

```diff
@@ -0,0 +1,323 @@
+			// call under test
+			new ReadLockRequest(mockCallback, context, keys);
+		}).getMessage();
+		assertEquals("ProgressCallback cannot be null", message);
+	}
+
+	@Test
+	public void testConstructorWithTimeoutTooLow() {
+		when(mockCallback.getLockTimeoutSeconds()).thenReturn(1L);
+		String message = assertThrows(IllegalArgumentException.class, () -> {
+			// call under test
+			new ReadLockRequest(mockCallback, context, keys);
+		}).getMessage();
+		assertEquals("LockTimeout cannot be less than 2 seconds", message);
+	}
+
+	@Test
+	public void testConstructorWithNullContext() {
+		when(mockCallback.getLockTimeoutSeconds()).thenReturn(maxTimeout);
+		context = null;
+		String message = assertThrows(IllegalArgumentException.class, () -> {
+			// call under test
+			new ReadLockRequest(mockCallback, context, keys);
+		}).getMessage();
+		assertEquals("Caller's context cannot be null", message);
+	}
+
+	@Test
+	public void testConstructorWithNullLockKeys() {
+		when(mockCallback.getLockTimeoutSeconds()).thenReturn(maxTimeout);
+		keys = null;
+		String message = assertThrows(IllegalArgumentException.class, () -> {
+			// call under test
+			new ReadLockRequest(mockCallback, context, keys);
+		}).getMessage();
+		assertEquals("LockKey cannot be null", message);
+	}
+
+	@Test
+	public void testConstructorWithEmptyLockKeys() {
+		when(mockCallback.getLockTimeoutSeconds()).thenReturn(maxTimeout);
+		keys = new String[0];
+		String message = assertThrows(IllegalArgumentException.class, () -> {
+			// call under test
+			new ReadLockRequest(mockCallback, context, keys);
+		}).getMessage();
+		assertEquals("Must include at least one lock key", message);
+	}
+
+	@Test
+	public void testConstructorWithNullInLockKeys() {
+		when(mockCallback.getLockTimeoutSeconds()).thenReturn(maxTimeout);
+		keys = new String[] { "one", null, "two" };
+		String message = assertThrows(IllegalArgumentException.class, () -> {
+			// call under test
+			new ReadLockRequest(mockCallback, context, keys);
+		}).getMessage();
+		assertEquals("Lock key cannot be null", message);
+	}
+
+	@Test
+	public void testConstructorWithNullSemaphore() {
+		when(mockCallback.getLockTimeoutSeconds()).thenReturn(maxTimeout);
+		mockCountingSemaphore = null;
+		String message = assertThrows(IllegalArgumentException.class, () -> {
+			// call under test
+			new ReadLockImpl(mockCountingSemaphore, maxNumberOfReaders, new ReadLockRequest(mockCallback, context, keys));
+		}).getMessage();
+		assertEquals("CountingSemaphore cannot be null", message);
+	}
+	
+	@Test
+	public void testConstructorWithNullRequest() {
+		String message = assertThrows(IllegalArgumentException.class, () -> {
+			// call under test
+			new ReadLockImpl(mockCountingSemaphore, maxNumberOfReaders, null);
+		}).getMessage();
+		assertEquals("ReadLockRequest cannot be null", message);
+	}
+
+	@Test
+	public void testAcqurieLockAndClose() throws LockUnavilableException, Exception {
```

Comment:
maybe separety the acquire vs close tests (the try with resource is already tested?